src/java/org/tensorics/core/tree/walking/Trees.java

```diff
@@ -197,16 +174,16 @@ public void onEvery(Node node) {
         return newRebuildingContext.getUpdatedOrSame(rootNode);
     }
 
-    /**
+	/**
      * walks through the tree, starting from the given rootNode and collects all the nodes which implement the given
      * class.
      * 
      * @param rootNode the node from which to start the search
      * @param nodeClassToFind the class of the nodes to find
      * @return a set of all found nodes, which implement the given class
      */
-    public static <T extends Node> Set<T> findNodesOfClass(Node rootNode, final Class<T> nodeClassToFind) {
-        final Set<T> foundNodes = new HashSet<>();
+    public static <T extends Node, C> Set<C> findNodesOfClass(T rootNode, final Class<C> nodeClassToFind) {

```

Comment:
Oh, I see ... sure, C could be a marker interface which does not extend Node, and the actual implementation then implements both C and Node. I missed that - fine for me!